### DISCUSSION TITLE
How to serialize an integer as hexadecimal or binary?

### DISCUSSION THREAD

**Discussion by pastelmind** [toml-rs/toml#781] [Q&A] [answered]
I want to serialize certain integer fields of a struct as hex or binary format.

The following struct...

```rust
#[derive(Serialize, Deserialize)]
struct Foo {
  dec: u32,
  hex: u32,
  bin: u32,
}

let foo = Foo { dec: 12, hex: 34, bin: 56 };
```

...would be serialized to:

```toml
dec = 12
hex = 0x22
bin = 0b111000
```

Is this possible?

**Comment by epage** [comment]
FYI this is more of serde question than toml. Check out the field attributes documented at serde.rs

**Reply by pastelmind** [reply]
But serde doesn't concern itself with how integers are converted to strings, does it? Serde operates on a higher level.

I'm not interested in what [serde-hex](https://github.com/fspmarshall/serde-hex) does. serde-hex converts integers into string literals:

```toml
[my_struct]
hex_field = "0x00010203"
```

But I want to exercise TOML's native support for hexadecimal and binary integers:

```toml
[my_struct]
hex_field = 0x00010203
```

I examined toml_edit and it appears that all integers (`i32`, `u8`, etc.) are converted to `i64` during serialization...

https://github.com/toml-rs/toml/blob/1bfb7d7d79880289fc1f10ef3f6dd0b2b3267c17/crates/toml_edit/src/ser/value.rs#L76-L109

...which is encoded as a base-10 integer here:

https://github.com/toml-rs/toml/blob/1bfb7d7d79880289fc1f10ef3f6dd0b2b3267c17/crates/toml_edit/src/encode.rs#L516-L520

Would it be possible to customize this behavior today? Or would I have to write a custom `Visit` implementation?

**Reply by epage** [reply, marked as answer]
Sorry, I missed that you were referring to how TOML renders the number.

serde does not natively pass through custom formatting.  Our options
- Have a special wrapper type, like we do with `serde_spanned`.  These are a pain to work with
- Make it a setting on the deserializer that applies to all integers.  This feels too niche to support.
- Deserialzie to a `toml_edit::Document` and apply your own visitor.

I currently suggest the last one when people want to do custom formatting.

That leaves us with where you left off, a way to control what repr gets generated for numbers.  The only way to do it today is for you to generate a valid TOML representation, parse it, and put it in the document.  This is not ideal.  For integers, the two main aspects we'd need to control are the base and the digit separators.  Our API options are
- Separate APIs for each.  When applying botth, we'd either need to store the format style (would rather not do this) or parse the repr to see what was done and re-generate
- A single format call for integers, taking in a non-exhaustive `IntegerFormat`
- A single format call for any scalar, taking in a non-exhaustive `ScalarFormat`.  This could then store all the settings for string formatting (see #590), etc

Not too sure what approach I'd like to be taking.

**Reply by pastelmind** [reply]
Thank you. I feel underqualified to provide an opinion on the matter at the moment. For now, I suppose a custom visitor would have to do.

**Reply by pastelmind** [reply]
I am revisiting this. I believe that:

- The ability to emit integers in hexadecimal, binary, or (perhaps less frequently) octal would be relatively easy to implement, since there is a small number of well-defined options. (I am ignoring hexadeximals with mixed case (e.g.`0xAbCdEf`) hexadecimals since they are almost always entirely uppercase or lowercase.)
- The ability to customize decimal separators would be more involved since there is a very large space of possible representations (e.g. `1_000_000` vs `100_0000` vs `1_0_0_0_0_0_0`).

So we could go for the low-hanging fruit.

I am working on a pull request (no guarantees for the delivery date!) that introduces an `Integer` type. I'm going with:

- The underlying numeric type is `i64`.
- The `Integer` type should have methods to read or mutate its value and notation in-place, as well as constructors for each notation.
- Constructing or converting to a hexadecimal, octal, or binary number fails if the value is negative, because the TOML spec does not allow them.
- There should be a fail-safe variant of the value setter that also converts the notation to decimal if the new value is negative.

Feedback is welcome!
